src/5.0_to_5.1/calculated-measures/migrateCalculatedMeasuresInWidgets.ts

```diff
@@ -0,0 +1,89 @@
+  WidgetWithQueryState,
+} from "@activeviam/activeui-sdk-5.0";
+import { migrateCalculatedMeasuresInMdx } from "./migrateCalculatedMeasuresInMdx";
+import { produce } from "immer";
+import _mapValues from "lodash/mapValues";
+import { serializeWidgetState } from "@activeviam/activeui-sdk-5.1";
+import _uniq from "lodash/uniq";
+
+/**
+ * In ActiveUI 5.0, saved calculated measures live under /ui/calculated_measures, and ActivePivot is not aware of them. In particular, they don't appear in the data model, and they must be added as query-scoped measures to each individual widget refering to them.
+ *
+ * In ActiveUI 5.1, saved calculated measures are grouped by cube and live under /pivot/entitlements/cm. ActivePivot proxies them and makes them accessible in the data model. So they can be referred to just like any other measure in an MDX query, without the need to add a query-scoped calculated measure definition.
+ */
+export const migrateCalculatedMeasuresInWidgets = (
+  widgets: ContentRecord,
+  dataModel: DataModel,
+  namesOfCalculatedMeasurestoMigrate: string[],
+): {
+  measureToCubeMapping: { [measureName: string]: CubeName[] };
+  migratedWidgetsRecord: ContentRecord;
+} => {
+  // Create an empty object where each calculated measure used in a saved widget will be added as a key with its cubeName as a value.
+  const measureToCubeMapping: { [measureName: string]: CubeName[] } = {};
+
+  const migratedWidgetsRecord = produce(widgets, (draft) => {
+    draft.children!.content.children = _mapValues(
+      draft.children!.content.children,
+      (widgetRecord) => {
+        const serializedWidgetState: WidgetWithQueryState<
+          MdxSelect,
+          "serialized"
+        > = JSON.parse(widgetRecord.entry.content);
+        const deserializedWidgetState = deserializeWidgetState(
+          serializedWidgetState,
+        );
+        if (isWidgetWithQueryState(deserializedWidgetState)) {
+          const mdx: MdxSelect | undefined = deserializedWidgetState.query.mdx;
+
+          if (!mdx) {
+            return widgetRecord;
+          }
+
+          const {
+            migratedMdx,
+            namesOfCalculatedMeasuresToMigrateInWidget,
+            cubeName,
+          } = migrateCalculatedMeasuresInMdx(
+            mdx,
+            namesOfCalculatedMeasurestoMigrate,
+            dataModel,
+          );
+
+          namesOfCalculatedMeasuresToMigrateInWidget.forEach(
+            (calculatedMeasureName) => {
+              measureToCubeMapping[calculatedMeasureName] =
+                measureToCubeMapping[calculatedMeasureName]
+                  ? _uniq([
+                      ...measureToCubeMapping[calculatedMeasureName],
+                      cubeName,
+                    ])
+                  : [cubeName];
+            },
+          );
+          const updatedWidgetState = produce(
+            deserializedWidgetState,
+            (draft) => {
+              draft.query.mdx = migratedMdx;
+            },
+          );
```

Comment:
No need to use `produce` here since there is already one above:
```suggestion
        deserializedWidgetState.query.mdx = migratedMdx;
```

Reply:
Making this change is giving me `TypeError: Cannot assign to read only property 'mdx' of object`. Do you have a suggestion how I could fix this error @antoinetissier ? 

![Capture d’écran 2022-12-02 113141](https://user-images.githubusercontent.com/64020381/205273075-722569e0-c256-4c11-b1b4-80b8f53aec81.png)